cli/bptest/metadata_lint.go

```diff
@@ -0,0 +1,126 @@
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"regexp"
+
+	"gopkg.in/yaml.v2"
+)
+
+// BlueprintConnection structure to match metadata YAML
+type BlueprintConnection struct {
+	Source  string `yaml:"source"`
+	Version string `yaml:"version"`
+}
+
+// BlueprintVariable structure to match metadata YAML
+type BlueprintVariable struct {
+	Connections []BlueprintConnection `yaml:"connections"`
+}
+
+// BlueprintInterface structure to match metadata YAML
+type BlueprintInterface struct {
+	Variables []BlueprintVariable `yaml:"variables"`
+}
+
+// BlueprintMetadata structure for entire metadata YAML
+type BlueprintMetadata struct {
+	Spec struct {
+		Interfaces []BlueprintInterface `yaml:"interfaces"`
+	} `yaml:"spec"`
+}
+
+// validateMetadataFile validates a metadata.yaml file by checking blueprint references and versions
+func validateMetadataFile(metadataPath string) error {
+	fmt.Printf("Validating: %s\n", metadataPath)
+
+	// Read the metadata file
+	content, err := ioutil.ReadFile(metadataPath)
+	if err != nil {
+		return fmt.Errorf("error reading metadata file %s: %v", metadataPath, err)
+	}
+
+	// Unmarshal the YAML content into the BlueprintMetadata struct
+	var metadata BlueprintMetadata
+	if err := yaml.Unmarshal(content, &metadata); err != nil {
+		return fmt.Errorf("error parsing metadata file %s: %v", metadataPath, err)
+	}
+
+	// Walk through the BlueprintInterface -> BlueprintVariable -> BlueprintConnection
+	for _, blueprintInterface := range metadata.Spec.Interfaces {
+		for _, blueprintVariable := range blueprintInterface.Variables {
+			for _, connection := range blueprintVariable.Connections {
+				// Validate blueprint source and version
+				if err := validateBlueprintSource(connection.Source); err != nil {
+					return err
+				}
+				if err := validateBlueprintVersion(connection.Version, connection.Source); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+var httpGetFunc = http.Get
+
+// validateBlueprintSource checks if the blueprint source exists
+func validateBlueprintSource(source string) error {
+	// Regex to validate source format
+	sourceRegex := regexp.MustCompile(`^[a-zA-Z0-9\-\.]+/[a-zA-Z0-9\-\.]+$`)
+	if !sourceRegex.MatchString(source) {
+		return fmt.Errorf("Invalid blueprint source format: %s. Expected format: namespace/repository", source)
+	}
+
+	// Check if the blueprint source exists.
+	resp, err := httpGetFunc(fmt.Sprintf("https://github.com/%s", source))
+	if err != nil || resp.StatusCode != 200 {
+		return fmt.Errorf("Blueprint source not found: %s", source)
+	}
+
+	fmt.Printf("Valid source: %s\n", source)
+	return nil
+}
+
+// validateBlueprintVersion checks if the given version constraint is valid and if the specified version exists.
+func validateBlueprintVersion(version string, source string) error {
+	// Regex for Terraform version constraints
+	versionConstraintRegex := regexp.MustCompile(`^((>=|<=|>|<|=|~>|\^)?\s?\d+(\.\d+){0,2})(,\s*(>=|<=|>|<|=|~>|\^)?\s?\d+(\.\d+){0,2})*$`)
```

Comment:
Let's use something like https://github.com/hashicorp/go-version rather than maintaining regexes which are hard to test.

Reply:
Done